### DISCUSSION TITLE
RequestClose: ObjectdisposedExceptions in Kestrel

### DISCUSSION THREAD

**Discussion by abko-auto** [dotnet/runtime#75788] [General]
When looking at perf issues came across ObjectDisposed exceptions from Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.KestrelConnection!RequestClose

And these ObjectDisposed exception are contributing ~10% of total exceptions. When looked at code KestrelConnection::RequestClose() There is comment about race condition - but this race condition is happening more than 10% of managed exceptions, can we somehow avoid these exceptions?

<img width="629" alt="image" src="https://user-images.githubusercontent.com/66053025/190827462-588441ec-afc8-41b4-a39f-9f09894ae989.png">

Entire stack
![image](https://user-images.githubusercontent.com/66053025/190827835-f623a626-35e8-4319-8ca9-def2d099ec9c.png)

**Comment by bartonjs** [comment]
"10% of your managed exceptions" doesn't mean "happens 10% of the time".  Presumably a lot of things result in no exceptions whatsoever?

**Reply by abko-auto** [reply]
Even if I collect perf trace its happening more than 10% of managed exception - doesnt mean 10% of the time but for us that 10% of exceptions in 30 seconds sampling was more than 200 exceptions.

**Reply by abko-auto** [reply]
Can we have more information about these ObjectDisposed exceptions, I see them all our services? is this something can we fix or we need to live with this? This is one of top 5 exception our application sees, and we would like to get rid of them.

**Comment by davidfowl** [comment]
Looks like you're calling RequestClose after the connection ends? It's being called from a timer thread and is likely unsynchronized with the connection ending.

**Reply by abko-auto** [reply]
Thanks @davidfowl for your response. Yes, we use your suggestion here to periodically close connections - https://github.com/dotnet/aspnetcore/issues/18893. Is there any better way to do that and avoid these objects disposed exceptions.

**Reply by davidfowl** [reply]
But from the stack it looks like it’s being called from a timer callback that isn’t taking into account the lifetime of the connection.

**Reply by abko-auto** [reply]
` options.Listen(
                                IPAddress.Any,
                                int.Parse(listenerPort),
                                opt => 
                                {
                                    if (connectionTimeoutEnabled && connectionTimeoutInSeconds > 0)
                                    {
                                        opt.Use(async (connection, next) =>
                                        {
                                            IConnectionLifetimeNotificationFeature connectionLifetimeNotification = connection.Features.Get<IConnectionLifetimeNotificationFeature>();
                                            if (connectionLifetimeNotification != null)
                                            {
                                                Task.Run(async () =>
                                                {
                                                    await Task.Delay(TimeSpan.FromSeconds(connectionTimeoutInSeconds));
                                                    connectionLifetimeNotification.RequestClose();
                                                });
                                            }

                                            await next();
                                        });
                                    }

                                    if (urlToUse.Contains("https", StringComparison.OrdinalIgnoreCase))
                                    {
                                        opt.UseHttps(certificate);
                                    }
                                });`


thats the code we use.

**Reply by davidfowl** [reply]
Yes this code is doing what I expected. You need to cancel the delay after the middleware has completed the middleware chain. There are 2 scenarios:
1. The middleware (next()) is running when you're about to run your delay.
2. The middleware has already completed when you're about to run your delay.

You're not handling the second case which is why your code is running after disposal. I can think of 3 ways to slice this. 

1. Something like what you have today but cancel the timeout when the middleware ends.

```C#
var cancelTimeoutSource = new CancellationTokenSource();

_ = Task.Run(async () =>
{
    await Task.Delay(connectionTimeout, cancelTimeoutSource.Token);
    context.Features.Get<IConnectionLifetimeNotificationFeature>()?.RequestClose();
});

try
{
    await next();
}
finally 
{
    // Cancel the timeout, it's no longer important since we completed
    cancelTimeoutSource.Cancel();
}
```

2. Using the cancellation token itself as the timer instead of `Task.Delay`.

```C#
var lifetime = context.Features.Get<IConnectionLifetimeNotificationFeature>();
if (lifetime is null)
{
    await next();
    return;
}

using var cancelTimeoutSource = new CancellationTokenSource();
using var reg = cancelTimeoutSource.Token.Register(static (state, ct) =>
{
    ((IConnectionLifetimeNotificationFeature)state!).RequestClose();
},
lifetime);
cancelTimeoutSource.CancelAfter(connectionTimeout);

await next();
```

3. Using the new WaitAsync method:

```C#
var lifetime = context.Features.Get<IConnectionLifetimeNotificationFeature>();
var task = next();

if (lifetime is null)
{
    await task;
    return;
}

try
{
    // Wait for the timeout to fire or the task to complete successfully
    await task.WaitAsync(connectionTimeout);
}
catch (TimeoutException)
{
    // The timeout fired so close the connection and wait for the task (it's all cooperative)
    lifetime.RequestClose();
    await task;
}
```

**Reply by abko-auto** [reply]
Thanks @davidfowl  for your quick responses and possible solutions.

**Reply by abko-auto** [reply]
@davidfowl - Isnt 2nd option suggested will have same problem - middleware is already finished and we are about to run registered delegate?

**Reply by davidfowl** [reply]
> @davidfowl - Isnt 2nd option suggested will have same problem - middleware is already finished and we are about to run registered delegate?

No the token will be disposed (see the `using var` syntax) before the middleware unwinds. As a result, it won't trigger the callback anymore, so the connection abort attempt won't happen.

**Reply by abko-auto** [reply]
Got it! Was reading the code as plain text. Considering its running as middleware - totally makes sense! 
Is this the pattern people use to close connections periodically? What we are trying to do is after every 5 minutes, sending RequestClose() back to the client to avoid sticky connections and balance our traffic evenly.

**Reply by davidfowl** [reply]
@abko-auto I don't know if people use this pattern. I haven't seen anyone a majority of people ask for it. We have a keep alive timeout by default that can take care of this if you are worried about idle connections. Also if you do have long lived connections, doing it at this level makes it hard to know which connections to timeout because there are only short lived requests on the connection vs a long lived streaming style connection (websocket, server sent events etc).

**Comment by abko-auto** [comment]
Exactly, after reading existing code and solution you mentioned above realized we are not doing whats expected. We do have services where  we expect long lived connection, do we have better solution to close long lived connections after certain time gracefully?

**Reply by davidfowl** [reply]
I think this pattern is still fine, the only other thing you'd want to add here potentially is the ability for a request to cancel the timeout. That way, code further up the stack that is results in a long running request/connection can cancel the timeout.
